asmcomp/cmm_helpers.ml

```diff
@@ -2215,29 +2212,28 @@ type binary_primitive = expression -> expression -> Debuginfo.t -> expression
 
 (* Helper for compilation of initialization and assignment operations *)
 
-type assignment_kind = Caml_modify | Caml_initialize | Simple
+type assignment_kind = Caml_modify | Simple
 
 let assignment_kind
     (ptr: Lambda.immediate_or_pointer)
     (init: Lambda.initialization_or_assignment) =
   match init, ptr with
   | Assignment, Pointer -> Caml_modify
-  | Heap_initialization, Pointer -> Caml_initialize
+  | Heap_initialization, _ ->
+     Misc.fatal_error "Cmm_helpers: Lambda.Heap_initialization unsupported"
   | Assignment, Immediate
-  | Heap_initialization, Immediate
   | Root_initialization, (Immediate | Pointer) -> Simple
 
 let setfield n ptr init arg1 arg2 dbg =
   match assignment_kind ptr init with
   | Caml_modify ->
+      (* In principle, we could track whether the block being modified
+         is possibly local and use caml_modify when it's not. However,
+         the extra complexity only saves a single branch, so for now
+         this unconditionally uses caml_modify_local *)
       return_unit dbg
-        (Cop(Cextcall("caml_modify", typ_void, [], false),
```

Comment:
There is also a `"caml_modify"` in `addr_array_set` which I think needs to be updated too, since I think we allow local arrays, or at least intend to.

Reply:
Currently, it's correct but a bit useless. You can make local arrays, but the typing of the array-set primitive means that only global arrays can be mutated, so it's correct to do caml_modify from addr_array_set.  I'll fix this in a future PR that adds local array primitives (which depends on #43).